rewrite-core/src/main/java/org/openrewrite/Parser.java

```diff
@@ -165,6 +177,46 @@ public boolean isSynthetic() {
             return synthetic;
         }
 
+        public Charset getCharacterEncoding() {
+            return characterEncoding;
+        }
+
+        /**
+         * Get the file encoding or default to UTF-8.
+         */
+        public static Charset detectCharacterEncoding(Path input) {
+            UniversalDetector detector = new UniversalDetector();
+            try (FileInputStream fis = new FileInputStream(input.toFile())) {
+                byte[] buf = new byte[4096];
+                int n;
+                while ((n = fis.read(buf)) > 0 && !detector.isDone()) {
+                    detector.handleData(buf, 0, n);
+                }
+                detector.dataEnd();
+            } catch (Exception ignored) {
+            }

```

Comment:
This library has a helper that takes a Path: 

https://github.com/albfernandez/juniversalchardet/blob/main/src/main/java/org/mozilla/universalchardet/UniversalDetector.java#L363-L367

I also like how this library will stop once it is confident enough in an answer.